Questions about DashscopeLLMClient Streaming ToolCall

When using DashscopeLLMClient to execute [StreamingAgentWithTools](https://github.com/JetBrains/koog/tree/develop/examples/simple-examples/src/main/kotlin/ai/koog/agents/example/streaming)，

<img width="915" height="1077" alt="Image" src="https://github.com/user-attachments/assets/6f28fd8d-d435-4e5c-aafb-041dd42f5b0d" />

An anomaly has occurred:

`Caused by: kotlinx.serialization.json.internal.JsonDecodingException: Unexpected JSON token at offset 0: Cannot read Json element because of unexpected end of the object '}' at path: $
JSON input: }`

After investigation, two potential issues were found in the code:

1. When the returned id is "", appendArgumentsDelta should be executed.but it is not.

  <img width="1679" height="843" alt="Image" src="https://github.com/user-attachments/assets/e1b89cdc-0b41-496a-ad73-f4cab3d1f5ff" />

<img width="1352" height="683" alt="Image" src="https://github.com/user-attachments/assets/5b28e9fc-9d0a-473f-b20a-eab099087f7c" />


  
2. When function.arguments is null, the generated JSON has issues.

<img width="1749" height="880" alt="Image" src="https://github.com/user-attachments/assets/bd81ff4d-d098-4277-a162-42a1bddb79c2" />

<img width="993" height="267" alt="Image" src="https://github.com/user-attachments/assets/82690ea3-2403-4ef4-a959-775795c9a74e" />


================================================================================
Since DashscopeLLMClient is final, it cannot be inherited.The only option is to write a similar "repeated" client, mimicking DashscopeLLMClient.  


I am puzzled, is there an issue with the data format returned by aliyun? Or do we need to adjust upsertToolCall? Thanks:)

--- a/prompt/prompt-executor/prompt-executor-clients/prompt-executor-dashscope-client/src/commonMain/kotlin/ai/koog/prompt/executor/clients/dashscope/DashscopeLLMClient.kt
+++ b/prompt/prompt-executor/prompt-executor-clients/prompt-executor-dashscope-client/src/commonMain/kotlin/ai/koog/prompt/executor/clients/dashscope/DashscopeLLMClient.kt
@@ -132,7 +132,7 @@ public class DashscopeLLMClient @JvmOverloads constructor(
                 choice.delta.content?.let { emitTextDelta(it) }
 
                 choice.delta.toolCalls?.forEach { toolCall ->
-                    val id = toolCall.id
+                    val id = toolCall.id?.takeIf { it.isNotEmpty() }
                     val name = toolCall.function?.name
                     val arguments = toolCall.function?.arguments
                     val index = toolCall.index

--- a/prompt/prompt-executor/prompt-executor-clients/prompt-executor-dashscope-client/src/jvmTest/kotlin/ai/koog/prompt/executor/clients/dashscope/models/DashscopeSerializationTest.kt
+++ b/prompt/prompt-executor/prompt-executor-clients/prompt-executor-dashscope-client/src/jvmTest/kotlin/ai/koog/prompt/executor/clients/dashscope/models/DashscopeSerializationTest.kt
@@ -436,6 +436,51 @@ class DashscopeSerializationTest {
         }
 
     @Test
+    fun `test stream response deserialization with null type and id in tool call chunks`() =
+        runWithBothJsonConfigurations("stream response with null type and id in tool call chunks") { json ->
+            val jsonInput =
+                // language=json
+                """
+            {
+                "id": "chatcmpl-tool-1",
+                "object": "chat.completion.chunk",
+                "created": 1677652600,
+                "model": "qwen-plus",
+                "choices": [
+                    {
+                        "index": 0,
+                        "delta": {
+                            "toolCalls": [
+                                {
+                                    "index": 0,
+                                    "id": null,
+                                    "type": null,
+                                    "function": {
+                                        "name": null,
+                                        "arguments": "{\"partial\":"
+                                    }
+                                }
+                            ]
+                        }
+                    }
+                ]
+            }
+                """.trimIndent()
+
+            val response = json.decodeFromString(DashscopeChatCompletionStreamResponse.serializer(), jsonInput)
+
+            response.id shouldBe "chatcmpl-tool-1"
+            response.choices.size shouldBe 1
+            val toolCalls = response.choices[0].delta.toolCalls
+            kotlin.test.assertNotNull(toolCalls)
+            toolCalls.size shouldBe 1
+            toolCalls[0].index shouldBe 0
+            toolCalls[0].id shouldBe null
+            toolCalls[0].type shouldBe null
+            toolCalls[0].function?.arguments shouldBe "{\"partial\":"
+        }
+
+    @Test
     fun `test deserialization of extended parameters`() =
         runWithBothJsonConfigurations("deserialization of extended parameters") { json ->
             val jsonInput =

--- a/prompt/prompt-executor/prompt-executor-clients/prompt-executor-openai-client-base/src/commonMain/kotlin/ai/koog/prompt/executor/clients/openai/base/models/OpenAIDataModels.kt
+++ b/prompt/prompt-executor/prompt-executor-clients/prompt-executor-openai-client-base/src/commonMain/kotlin/ai/koog/prompt/executor/clients/openai/base/models/OpenAIDataModels.kt
@@ -302,7 +302,7 @@ public class OpenAIToolCall(
  * @property id The unique identifier associated with the tool call. This value can be null.
  * @property function The function object containing the name and arguments of the function invoked.
  * This value can be null if no function call is associated.
- * @property type The type of the tool call. This is always set to "function" for denoting function-based calls.
+ * @property type The type of the tool call. Defaults to "function" for denoting function-based calls. Can be null.
  */
 @Serializable
 public class OpenAIStreamToolCall(

--- a/prompt/prompt-executor/prompt-executor-clients/prompt-executor-openai-client-base/src/commonMain/kotlin/ai/koog/prompt/executor/clients/openai/base/models/OpenAIDataModels.kt
+++ b/prompt/prompt-executor/prompt-executor-clients/prompt-executor-openai-client-base/src/commonMain/kotlin/ai/koog/prompt/executor/clients/openai/base/models/OpenAIDataModels.kt
@@ -308,11 +308,9 @@ public class OpenAIToolCall(
 public class OpenAIStreamToolCall(
     public val index: Int,
     public val id: String?,
-    public val function: OpenAIStreamFunction?
-) {
-    /** The type of the tool. Currently, only `function` is supported. */
-    public val type: String = "function"
-}
+    public val function: OpenAIStreamFunction?,
+    public val type: String? = "function"
+)
 
 /**
  * Function call from an OpenAI model, containing the function name and arguments.

--- a/prompt/prompt-model/src/commonMain/kotlin/ai/koog/prompt/streaming/StreamFrameFlowBuilder.kt
+++ b/prompt/prompt-model/src/commonMain/kotlin/ai/koog/prompt/streaming/StreamFrameFlowBuilder.kt
@@ -256,7 +256,6 @@ public class StreamFrameFlowBuilder(
         val index: Int?,
     ) {
         fun appendArgumentsDelta(argumentsDelta: String?): PendingToolCall {
-            require(this.index == index)
             val newArgs =
                 if (argumentsDelta == null) this.argumentsDelta else (this.argumentsDelta ?: "") + argumentsDelta
             return copy(argumentsDelta = newArgs)

--- a/prompt/prompt-model/src/commonMain/kotlin/ai/koog/prompt/streaming/StreamFrameFlowBuilder.kt
+++ b/prompt/prompt-model/src/commonMain/kotlin/ai/koog/prompt/streaming/StreamFrameFlowBuilder.kt
@@ -257,7 +257,9 @@ public class StreamFrameFlowBuilder(
     ) {
         fun appendArgumentsDelta(argumentsDelta: String?): PendingToolCall {
             require(this.index == index)
-            return copy(argumentsDelta = (this.argumentsDelta ?: "") + argumentsDelta)
+            val newArgs =
+                if (argumentsDelta == null) this.argumentsDelta else (this.argumentsDelta ?: "") + argumentsDelta
+            return copy(argumentsDelta = newArgs)
         }
     }
 
@@ -267,7 +269,8 @@ public class StreamFrameFlowBuilder(
     ) {
         fun appendTextDelta(textDelta: String?, index: Int?): PendingText {
             require(this.index == index)
-            return copy(textDelta = (this.textDelta ?: "") + textDelta)
+            val newText = if (textDelta == null) this.textDelta else (this.textDelta ?: "") + textDelta
+            return copy(textDelta = newText)
         }
     }
 

--- a/prompt/prompt-model/src/commonTest/kotlin/ai/koog/prompt/streaming/StreamFrameFlowBuilderTest.kt
+++ b/prompt/prompt-model/src/commonTest/kotlin/ai/koog/prompt/streaming/StreamFrameFlowBuilderTest.kt
@@ -232,6 +232,27 @@ class StreamFrameFlowBuilderTest {
     }
 
     @Test
+    fun testEmitToolCallDeltaWithNullArgumentsDoesNotCorruptContent() = runTest {
+        val frames = buildStreamFrameFlow {
+            emitToolCallDelta(id = "call_1", name = "search", args = "{\"query\":")
+            emitToolCallDelta(args = null)
+            emitToolCallDelta(args = "\"test\"}")
+            emitEnd()
+        }.toList()
+
+        assertContentEquals(
+            listOf(
+                StreamFrame.ToolCallDelta("call_1", "search", "{\"query\":"),
+                StreamFrame.ToolCallDelta(null, null, null),
+                StreamFrame.ToolCallDelta(null, null, "\"test\"}"),
+                StreamFrame.ToolCallComplete("call_1", "search", "{\"query\":\"test\"}"),
+                StreamFrame.End(null, ResponseMetaInfo.Empty)
+            ),
+            frames
+        )
+    }
+
+    @Test
     fun testEmitEndFlushesAllPendingFrames() = runTest {
         val frames = buildStreamFrameFlow {
             emitToolCallDelta(id = "call_1", name = "tool", args = "{}")

